### PR TITLE
Add support for \e escape sequences in c# strings and characters

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Language feature approved by the LDM for the any-time bucket: https://github.com/dotnet/csharplang/issues/7400

### CODE REVIEW

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/EmbeddedLanguages/VirtualChars/CSharpVirtualCharService.cs [hunk 1]
```diff
@@ -358,6 +358,7 @@ public override bool TryGetEscapeCharacter(VirtualChar ch, out char escapedChar)
                 case '0': ch = '\0'; break;
                 case 'a': ch = '\a'; break;
                 case 'b': ch = '\b'; break;
+                case 'e': ch = '\u001b'; break;
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/EmbeddedLanguages/VirtualChars/CSharpVirtualCharService.cs:361]: will add IDE tests as well.  But just wanted to get the ball rolling on this.

**Comment by jaredpar** [comment]
Don't forget to update language feature status.md

File: src/Compilers/CSharp/Test/Syntax/LexicalAndXml/LexicalTests.cs [hunk 2]
```diff
@@ -814,6 +815,61 @@ public void TestStringLiteralWithBadUnicodeEscape()
...
+
+        [Fact]
+        [Trait("Feature", "Literals")]
+        public void TestStringLiteralWithEscape_Preview()
+        {
+            var text = """
+                "\e"
+                """;
+            var value = "\u001b";
+            var token = LexToken(text, TestOptions.RegularPreview);
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Syntax/LexicalAndXml/LexicalTests.cs:845]: `RegularPreview` is the default. I think you intend to test using `RegularNext`, which will be updated mechanically to `Regular13` once it becomes available.

File: src/Compilers/CSharp/Test/Syntax/LexicalAndXml/LexicalErrorTests.cs [hunk 3]
```diff
@@ -128,6 +128,235 @@ public static void Main()
...
+                namespace x
+                {
+                    public class a
+                    {
+                        string a = "\e";
+                    }
+                }
+                """;
+
+            ParserErrorMessageTests.ParseAndValidate(test, TestOptions.RegularPreview);
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Test/Syntax/LexicalAndXml/LexicalErrorTests.cs:325]:
```suggestion
            ParserErrorMessageTests.ParseAndValidate(test, TestOptions.RegularNext);
```

File: src/Compilers/CSharp/Test/Syntax/LexicalAndXml/LexicalErrorTests.cs [hunk 4]
```diff
@@ -128,6 +128,235 @@ public static void Main()
...
+                namespace x
+                {
+                    public class a
+                    {
+                        string a = "e\ee";
+                    }
+                }
+                """;
+
+            ParserErrorMessageTests.ParseAndValidate(test, TestOptions.RegularPreview);
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Test/Syntax/LexicalAndXml/LexicalErrorTests.cs:341]:
```suggestion
            ParserErrorMessageTests.ParseAndValidate(test, TestOptions.RegularNext);
```

File: src/Compilers/CSharp/Test/Syntax/LexicalAndXml/LexicalErrorTests.cs [hunk 5]
```diff
@@ -128,6 +128,235 @@ public static void Main()
...
+                namespace x
+                {
+                    public class a
+                    {
+                        string a = '\e';
+                    }
+                }
+                """;
+
+            ParserErrorMessageTests.ParseAndValidate(test, TestOptions.RegularPreview);
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Test/Syntax/LexicalAndXml/LexicalErrorTests.cs:357]:
```suggestion
            ParserErrorMessageTests.ParseAndValidate(test, TestOptions.RegularNext);
```

File: src/Compilers/CSharp/Test/Syntax/LexicalAndXml/LexicalTests.cs (hunk 2 above)
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Test/Syntax/LexicalAndXml/LexicalTests.cs:845]:
```suggestion
            var token = LexToken(text, TestOptions.RegularNext);
```

File: src/Compilers/CSharp/Test/Syntax/LexicalAndXml/LexicalTests.cs [hunk 6]
```diff
@@ -942,6 +998,39 @@ public void TestCharacterLiteralEscape_V()
...
+            Assert.Equal(value, token.ValueText);
+        }
+
+        [Fact]
+        [Trait("Feature", "Literals")]
+        public void TestCharacterLiteralEscape_E_CSharpPreview()
+        {
+            var value = "\u001b";
+            var text = "'\\e'";
+            var token = LexToken(text, TestOptions.RegularPreview);
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Test/Syntax/LexicalAndXml/LexicalTests.cs:1024]:
```suggestion
            var token = LexToken(text, TestOptions.RegularNext);
```

File: src/Compilers/CSharp/Test/Syntax/LexicalAndXml/LexicalTests.cs (hunk 2 above)
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Test/Syntax/LexicalAndXml/LexicalTests.cs:845]: done.  thanks!

File: docs/Language Feature Status.md [hunk 7]
```diff
@@ -14,6 +14,7 @@ efforts behind them.
...
 | [Default in deconstruction](https://github.com/dotnet/roslyn/pull/25562) | [decon-default](https://github.com/dotnet/roslyn/tree/features/decon-default) | [In Progress](https://github.com/dotnet/roslyn/issues/25559) | [jcouv](https://github.com/jcouv) | [gafter](https://github.com/gafter) |  | [jcouv](https://github.com/jcouv) |
 | [Roles/Extensions](https://github.com/dotnet/csharplang/issues/5497) | [roles](https://github.com/dotnet/roslyn/tree/features/roles) | [In Progress](https://github.com/dotnet/roslyn/issues/66722) | [jcouv](https://github.com/jcouv) | [AlekseyTs](https://github.com/AlekseyTs), [jjonescz](https://github.com/jjonescz) |  | [MadsTorgersen](https://github.com/MadsTorgersen) |
+| [Escape character](https://github.com/dotnet/csharplang/issues/7400) | N/A | [In Progress](https://github.com/dotnet/roslyn/pull/70497) | [CyrusNajmabadi](https://github.com/CyrusNajmabadi) | [CyrusNajmabadi](https://github.com/CyrusNajmabadi) |  | [CyrusNajmabadi](https://github.com/CyrusNajmabadi) |
```
Comment (jcouv) [inline docs/Language Feature Status.md:17]: >[CyrusNajmabadi](https://github.com/CyrusNajmabadi) [](http://example.com/codeflow?start=194&length=51)

Put whoever reviews as reviewers (me and someone else)

**Review by jcouv — APPROVED** [review]
Compiler side LGTM Thanks (iteration 6)

File: docs/Language Feature Status.md (hunk 7 above)
Comment (CyrusNajmabadi) [reply docs/Language Feature Status.md:17]: will do.  when i get a second reviewer :)

Comment (CyrusNajmabadi) [inline docs/Language Feature Status.md:17]:
```suggestion
| [Escape character](https://github.com/dotnet/csharplang/issues/7400) | N/A | [In Progress](https://github.com/dotnet/roslyn/pull/70497) | [CyrusNajmabadi](https://github.com/CyrusNajmabadi) | [jcouv](https://github.com/jcouv) |  | [CyrusNajmabadi](https://github.com/CyrusNajmabadi) |
```

**Comment by CyrusNajmabadi** [comment]
@dotnet/roslyn-compiler @jaredpar  for another set of eyes.

File: docs/Language Feature Status.md [hunk 8]
```diff
@@ -14,6 +14,7 @@ efforts behind them.
...
 | [Default in deconstruction](https://github.com/dotnet/roslyn/pull/25562) | [decon-default](https://github.com/dotnet/roslyn/tree/features/decon-default) | [In Progress](https://github.com/dotnet/roslyn/issues/25559) | [jcouv](https://github.com/jcouv) | [gafter](https://github.com/gafter) |  | [jcouv](https://github.com/jcouv) |
 | [Roles/Extensions](https://github.com/dotnet/csharplang/issues/5497) | [roles](https://github.com/dotnet/roslyn/tree/features/roles) | [In Progress](https://github.com/dotnet/roslyn/issues/66722) | [jcouv](https://github.com/jcouv) | [AlekseyTs](https://github.com/AlekseyTs), [jjonescz](https://github.com/jjonescz) |  | [MadsTorgersen](https://github.com/MadsTorgersen) |
+| [Escape character](https://github.com/dotnet/csharplang/issues/7400) | N/A | [In Progress](https://github.com/dotnet/roslyn/pull/70497) | [CyrusNajmabadi](https://github.com/CyrusNajmabadi) | [jcouv](https://github.com/jcouv) |  | [CyrusNajmabadi](https://github.com/CyrusNajmabadi) |
```
Comment (CyrusNajmabadi) [inline docs/Language Feature Status.md:17]:
```suggestion
| [Escape character](https://github.com/dotnet/csharplang/issues/7400) | N/A | [In Progress](https://github.com/dotnet/roslyn/pull/70497) | [CyrusNajmabadi](https://github.com/CyrusNajmabadi) | [jcouv](https://github.com/jcouv) |  | [CyrusNajmabadi](https://github.com/CyrusNajmabadi) | 
```

**Comment by CyrusNajmabadi** [comment]
@dotnet/roslyn-compiler  @RikkiGibson  can i get a set of eyes on this small change?  thanks!

**Review by RikkiGibson — APPROVED** [review]

File: docs/Language Feature Status.md [hunk 9]
```diff
@@ -14,6 +14,7 @@ efforts behind them.
...
```
Comment (CyrusNajmabadi) [inline docs/Language Feature Status.md:17]:
```suggestion
| [Escape character](https://github.com/dotnet/csharplang/issues/7400) | N/A | [In Progress](https://github.com/dotnet/roslyn/pull/70497) | [CyrusNajmabadi](https://github.com/CyrusNajmabadi) | [jcouv](https://github.com/jcouv), [RikkiGibson](https://github.com/RikkiGibson) |  | [CyrusNajmabadi](https://github.com/CyrusNajmabadi) | 
```

**Pull request merged**